metrics.go

```diff
@@ -0,0 +1,434 @@
+// limitations under the License.
+
+package stackdriver
+
+/*
+The code in this file is responsible for converting OpenCensus Proto metrics
+directly to Stackdriver Metrics.
+*/
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/golang/protobuf/ptypes/timestamp"
+	"go.opencensus.io/trace"
+
+	distributionpb "google.golang.org/genproto/googleapis/api/distribution"
+	labelpb "google.golang.org/genproto/googleapis/api/label"
+	googlemetricpb "google.golang.org/genproto/googleapis/api/metric"
+	monitoredrespb "google.golang.org/genproto/googleapis/api/monitoredres"
+	monitoringpb "google.golang.org/genproto/googleapis/monitoring/v3"
+
+	"go.opencensus.io/metric/metricdata"
+	"go.opencensus.io/resource"
+)
+
+var (
+	errLableExtraction       = errors.New("error extracting labels")
+	errUnspecifiedMetricKind = errors.New("metric kind is unpsecified")
+)
+
+// ExportMetrics exports OpenCensus Metrics to Stackdriver Monitoring.
+func (se *statsExporter) ExportMetrics(ctx context.Context, metrics []*metricdata.Metric) error {
+	if len(metrics) == 0 {
+		return nil
+	}
+
+	for _, metric := range metrics {
+		se.protoMetricsBundler.Add(metric, 1)
+		// TODO: [rghetia] handle errors.
+	}
+
+	return nil
+}
+
+func (se *statsExporter) handleMetricsUpload(metrics []*metricdata.Metric) {
+	err := se.uploadMetrics(metrics)
+	if err != nil {
+		se.o.handleError(err)
+	}
+}
+
+func (se *statsExporter) uploadMetrics(metrics []*metricdata.Metric) error {
+	ctx, cancel := se.o.newContextWithTimeout()
+	defer cancel()
+
+	ctx, span := trace.StartSpan(
+		ctx,
+		"contrib.go.opencensus.io/exporter/stackdriver.uploadMetrics",
+		trace.WithSampler(trace.NeverSample()),
+	)
+	defer span.End()
+
+	for _, metric := range metrics {
+		// Now create the metric descriptor remotely.
+		if err := se.createMetricDescriptorFromMetric(ctx, metric); err != nil {
+			span.SetStatus(trace.Status{Code: 2, Message: err.Error()})
+			return err
+		}
+	}
+
+	var allTimeSeries []*monitoringpb.TimeSeries
+	for _, metric := range metrics {
+		tsl, err := se.metricToMpbTs(ctx, metric)
+		if err != nil {
+			span.SetStatus(trace.Status{Code: 2, Message: err.Error()})
+			return err
+		}
+		if tsl != nil {
+			allTimeSeries = append(allTimeSeries, tsl...)
+		}
+	}
+
+	// Now batch timeseries up and then export.
+	for start, end := 0, 0; start < len(allTimeSeries); start = end {
+		end = start + maxTimeSeriesPerUpload
+		if end > len(allTimeSeries) {
+			end = len(allTimeSeries)
+		}
+		batch := allTimeSeries[start:end]
+		ctsreql := se.combineTimeSeriesToCreateTimeSeriesRequest(batch)
+		for _, ctsreq := range ctsreql {
+			if err := createTimeSeries(ctx, se.c, ctsreq); err != nil {
+				span.SetStatus(trace.Status{Code: trace.StatusCodeUnknown, Message: err.Error()})
+				// TODO(@rghetia): Don't fail fast here, perhaps batch errors?
+				// return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// metricToMpbTs converts a metric into a Stackdriver Monitoring v3 API CreateTimeSeriesRequest
```

Comment:
s/a Stackdriver Monitoring v3 API CreateTimeSeriesRequest/a list of Stackdriver Monitoring v3 API TimeSeries

Reply:
changed.